### ISSUE TITLE
Implement DOM selections

### ISSUE THREAD

**Issue by jdm** [servo/servo#7492]
I propose that we store the bounds of our selection in `Document` using a starting node, an ending node, a starting offset, and an ending offset. Additionally, we should store a vector of these to enable multiple simultaneous selections in the future, and send all of this data along with each script-initiated reflow request. When a node is part of a selection, it should have a flag set on it (c.f. [NodeFlags](http://mxr.mozilla.org/servo/source/components/script/dom/node.rs#141)); when a node with this flag is encountered, the layout code should check whether it's the starting or ending node; if it is not, we can go ahead and style the node entirely as selected. If it is, we need to deal with styling the node content in up to three separate parts (pre-selection, in selection, post-selection).

cc @aweinstock314

**Comment by nox** [comment]
What if we use a DOM range for the selection boundaries?

**Comment by jdm** [comment]
Seems reasonable at first glance.

**Comment by jdm** [comment]
#4409 is related and I can never find it.

**Comment by jdm** [comment]
@iamrohit7 If you're interested in trying this, I will do my best to advise you!

**Comment by iamrohit7** [comment]
@jdm Thanks. I'll get back to you with any questions.

**Comment by pshaughn** [comment]
WPT custom-elements/reactions/Selection needs this.

**Comment by pshaughn** [comment]
Most failures under dom/ranges are also this.

**Comment by pshaughn** [comment]
Would implement DOM selections, up to a point that satisfies non-webdriver automated tests, be a productive thing to do in itself? I imagine they're not very useful without also having a UI to select things, but on the other hand that UI might be easier to write as its own piece on top of the already-working DOM part.

**Comment by jdm** [comment]
That would be a useful starting point, indeed.

**Comment by pshaughn** [comment]
Here's my understanding of what https://w3c.github.io/selection-api/ tells us to do.

Section 2:

- This is mostly preamble but it points out that Document.open resets the selection when it resets the document, which isn't spelled out anywhere in the meatier sections.

Section 3:

- Selection is a new DOM object, which can have references to a range and a couple nodes.
- It has many methods that modify or test the range; these are the bulk of the new code to add, but most of them are straightforward.
- It can return a live reference to the range and be affected by subsequent changes to it.

Section 4:

- Document gets a DomRefCell<Selection>, which is populated with an empty selection when the document's created.
- Document.getSelection returns the selection if there's a browsing context for the document, or null if not.
- Window.getSelection calls Document.getSelection.
- onselectstart and onselectionchange get added to the global event handlers (XRSession also uses the word onselectstart, but doesn't use a global event mixin so won't collide)

Section 5:

- When the existing selection is empty or collapsed, a selection UI action (e.g. drag) fires a selectstart event at the node the attempted new selection is starting on, and if this event is cancelled the selection isn't made. This is the only part of the spec that prescribes any UI functionality.
- Whenever the selection's range changes either by UI action or script, whether by setting a new range, discarding the existing range, or even a mutation to the existing range, a non-cancelable selectionchange event is fired at the document.

**Comment by pshaughn** [comment]
The selectionchange event has no tests and some underspecification (https://github.com/w3c/selection-api/issues/117) but I'll try to make some version of it work. Aside from that, I have a build that seems to be otherwise working, passing WPT tests that incidentally use getSelection, but I'm also getting some unexpected harness timeouts and I don't know how they relate to my changes yet. I'll push a draft of whatever I've got today, hopefully in a state where someone else could pick up the UI part.

**Comment by jdm** [comment]
In response to your [task source question](https://mozilla.logbot.info/servo/20200201#c16871375-c16871376), https://github.com/servo/servo/commit/7b3cf27c69f1ed2d5372e2f1e50ce2aab84e05a7#diff-a1afe92a339a20989be95f03c80740e5 is a nice example of adding a new task source.

**Comment by pshaughn** [comment]
The next big step here is figuring out how to convert a click into a range boundary, in the general case where we don't know in advance if it's going to be counting characters or counting child nodes.

**Comment by TimvdLippe** [comment]
@mrobinson While working on `document.execCommand`, I discovered a blocking bug: the `document.getSelection()` object is not shared across the document. Per https://w3c.github.io/selection-api/#dfn-selection

> This one [selection](https://w3c.github.io/selection-api/#dfn-selection) must be shared by all the content of the [document](https://dom.spec.whatwg.org/#document) (though not by nested [documents](https://dom.spec.whatwg.org/#document)), including any [editing hosts](https://html.spec.whatwg.org/multipage/interaction.html#editing-host) in the [document](https://dom.spec.whatwg.org/#document).

Currently, I call `document.GetSelection()` in `selection_if_command_is_enabled`, but it has an empty range. The tests call `textarea.selectionStart` and `textarea.selectionEnd` which updates the selection of the respective `textarea`, but that isn't reflected for the document itself.

I do see that between layout and script, we have a `SharedSelection`. That looks similar, but not completely the same.

When trying to figure out what's possible here, I naively put the following code in `HTMLTextAreaElement::maybe_update_shared_selection`:

```rust
let document = self.owner_document();
let can_gc = CanGc::note();
if let Some(selection) = document.GetSelection(can_gc) {
    selection.AddRange(&Range::new(
        &document,
        self.upcast(),
        start as u32,
        self.upcast(),
        end as u32,
        can_gc,
    ));
}
```

While it does compile, I don't think this is the right way to do it and it also crashes since `start` is a rope index that I don't know how to translate to a range offset.

What would be the best way forward here?

**Comment by pshaughn** [comment]
I think I might remember this. If I am recalling it correctly, it seemed to me at the time that the notion of a textarea having a selection was _distinct_ from the document having a selection, though I don't know whether that was because the specifications were different at the time or because I hadn't read them properly. I believe my changes to add a document selection didn't involve Servo's textarea code, and if the selection does need to be unified in this way, the textarea code will need to be changed to make textarea selection actions affect the document selection.

**Comment by TimvdLippe** [comment]
Okay, did some more digging and the `textarea` isn't actually the issue. The issue is that for `contenteditable` containers, we don't update the selection. That's tested in https://wpt.fyi/results/selection/contenteditable?product=servo

It makes sense that this isn't working yet, since we only recently started implementation of `contenteditable`. So I will tackle these tests first, before I can move on with `document.execCommand`.

**Comment by mrobinson** [comment]
> I think I might remember this. If I am recalling it correctly, it seemed to me at the time that the notion of a textarea having a selection was _distinct_ from the document having a selection, though I don't know whether that was because the specifications were different at the time or because I hadn't read them properly. I believe my changes to add a document selection didn't involve Servo's textarea code, and if the selection does need to be unified in this way, the textarea code will need to be changed to make textarea selection actions affect the document selection.

Yes, this can be observed in Chromium or Firefox by selecting text in a `<textarea>` and then selecting text outside the text area. If you then tab back to the `<textarea>`, the original selection is restored. I believe that the way this works is that that the DOM selection API should operate on whatever selection is active at the moment, but still the input field selection should coexist with the page selection when it doesn't have focus. This should all be confirmed though.

**Comment by mrobinson** [comment]
I think that we can close this now. Almost all of the DOM-side of selections is now implemented. There are refinements we can make, but, in general, most things work.
